pkg/virt-api/rest/expand.go

```diff
@@ -64,20 +75,83 @@ func (app *SubresourceAPIApp) ExpandSpecVMRequestHandler(request *restful.Reques
 		return
 	}
 
-	expandSpecResponse(vm, app.instancetypeMethods, errors.NewInternalError, response)
+	app.expandSpecResponse(vm, request, response, k8serr.NewInternalError)
 }
 
-func expandSpecResponse(vm *v1.VirtualMachine, instancetypeMethods instancetype.Methods, errorFunc func(error) *errors.StatusError, response *restful.Response) {
-	instancetypeSpec, err := instancetypeMethods.FindInstancetypeSpec(vm)
+func (app *SubresourceAPIApp) expandSpecResponse(vm *v1.VirtualMachine, request *restful.Request, response *restful.Response, errorFunc func(error) *k8serr.StatusError) {
+	authclient := app.virtCli.AuthorizationV1().SubjectAccessReviews()
+	authorizer := NewAuthorizorFromClient(authclient)
+	sar, err := authorizer.NewSubjectAccessReview(request)
+	if err != nil {
+		writeError(k8serr.NewInternalError(err), response)
+		return
+	}
+
+	instancetypeSpec, err := app.instancetypeMethods.FindInstancetypeSpec(vm)
 	if err != nil {
 		writeError(errorFunc(err), response)
 		return
 	}
-	preferenceSpec, err := instancetypeMethods.FindPreferenceSpec(vm)
+
+	instancetypeResourceAttributes, err := instancetype.CreateInstancetypeResourceAttributes(vm, "get")
```

Comment:
Do you think it makes sense to move this line before calling `FindInstancetypeSpec()`?
It seems unusual to me that permission to access instancetype is checked after accessing the instancetype.

Reply:
The order was intentionally swapped to prevent authorization checks on non-existent objects. WDYT?

Reply:
I'm not sure. I will leave the decision to the approver.

Reply:
/lgtm cancel

Thinking about this some more wouldn't this ordering potentially leak the fact that a resource existed or not without the user necessarily having the correct privileges to see it?

Appreciate this is just instancetypes and preferences so leaking that to the caller might not be the end of the world but reordering to ensure they have access before checking for the resource might be cleaner.

Reply:
Fair enough, you are right, it could leak the fact a resource doesn't exist. I swapped the order again, so authorization checks come first.